cirq/experiments/t2_decay_experiment.py

```diff
@@ -155,6 +170,28 @@ def t2_decay(
             study.Points('inv_y', [-0.5, 0.0]),
         )
         sweep = study.Product(delay_sweep, tomography_sweep)
+    elif experiment_type == ExperimentType.CPMG:
+        # Carr-Purcell-Meiboom-Gill sequence.
+        # Performs the following sequence
+        # π/2 - wait(t) - π - wait(2t) - ... - π - wait(t)
+        # There will be N π pulses (X gates)
+        # where N sweeps over the values of num_pulses
+        #
+        if num_pulses is None:
+            raise ValueError('At least one value must be given '
+                             'for num_pulses in a CPMG experiment')
+        max_pulses = max(num_pulses)
+        circuit = _cpmg_circuit(qubit, delay_var, max_pulses)
+        circuit.append(ops.X(qubit)**inv_x_var)
+        circuit.append(ops.Y(qubit)**inv_y_var)
+        circuit.append(ops.measure(qubit, key='output'))
+        tomography_sweep = study.Zip(
+            study.Points('inv_x', [0.0, 0.5]),
+            study.Points('inv_y', [-0.5, 0.0]),
+        )
+        pulse_sweep = _cpmg_sweep(num_pulses)
+        sweep = study.Product(delay_sweep, pulse_sweep, tomography_sweep)
+
```

Comment:
The experiments are very similar to each other and consequently there is a lot of repetition here. For example, `tomography_sweep` is defined three times.

Also, something that makes conceptual sense and will help reduce repetition is the separation of T2 and tomography circuits (the latter is exactly the same in each experiment).

I think it should be possible to have only one block of code to handle all three experiments in a readable way.

Reply:
I combined the echo code and CPMG.  I am not sure if it is worth the hit in readability to try to combine ramsey as well.  I think the preference would be to have a bit of verboseness since this is basically an extended example.

Reply:
Yes, leaving Ramsey circuit on its own is ok.

That said, you should move `tomography_sweep` out of the conditionals since it is literally identical for all experiments.

You should also move `sweep` outside the experiment-type-dependent blocks since the initialization you do for CPMG will work for Ramsey too (with a minor tweak to the condition on max_pulses).